IPython/core/display.py

```diff
@@ -140,6 +141,88 @@ def _new_id():
     return b2a_hex(os.urandom(16)).decode('ascii')
 
 
+class _RecursiveObject:
+    """
+    Default recursive object that provides a recursion repr if needed.
+    """
+
+    def __repr__(self):
+        return '<recursion ...>'
+
+    def _repr_html_(self):
+        return '&lt;recursion ...&gt;'
+
+
+DataMetadata = namedtuple('DataMetadata', ['data', 'metadata'])
+
+_objs = set()
+
+
+
+def get_repr(obj, include=None, exclude=None, *, on_recursion=_RecursiveObject):
+    """
+    return the representations of an object and associated metadata.
+
+    An given object can have many representation available, that can be defined
+    in many ways: `_repr_*_` methods, `_repr_mimebundle_`, or user-registered
+    formatter for types.
+
+    When given an object, `get_repr()` will search for the various formatting
+    option with their associated priority and return the requested
+    representation and associated metadata.
+
+
+    Parameters
+    ----------
+    obj : an objects
+        The Python objects to get the representation data.
+    include : list, tuple or set, optional
+        A list of format type strings (MIME types) to include in the
+        format data dict. If this is set *only* the format types included
+        in this list will be computed.
+    exclude : list, tuple or set, optional
+        A list of format type strings (MIME types) to exclude in the format
+        data dict. If this is set all format types will be computed,
+        except for those included in this argument.
+    on_recursion: callable
+        Return an object to compute the representation when recursion is
+        detected.
+
+
+    Returns
+    -------
+    (format_dict, metadata_dict) : named tuple of two dicts
+
+        0/data: format_dict is a dictionary of key/value pairs, one of each
+        format that was generated for the object. The keys are the format
+        types, which will usually be MIME type strings and the values and
+        JSON'able data structure containing the raw data for the
+        representation in that format.
+
+        1/metadata: metadata_dict is a dictionary of metadata about each
+        mime-type output. Its keys will be a strict subset of the keys in
+        format_dict.
+
+    Note
+    ----
+
+    When `get_repr` detect it is recursively called, it will attempt to return
+    the representation of `_RecursiveObject`
+
+    """
+    global _objs
```

Comment:
I'm not keen on the use of a global store here - you could get weird results if there's more than one thread or coroutine using this function at once. I'll try to think of a better way to handle the recursion thing.